metadata-ingestion/src/datahub/ingestion/source_report/ingestion_stage.py

```diff
@@ -25,25 +26,52 @@ class IngestionStageReport:
     ingestion_stage: Optional[str] = None
     ingestion_stage_durations: TopKDict[str, float] = field(default_factory=TopKDict)
 
-    _timer: PerfTimer = PerfTimer()
+    _timer: Optional[PerfTimer] = field(
+        default=None, init=False, repr=False, compare=False
+    )
 
-    def _close_stage(self) -> None:
-        if self._timer.is_running():
-            elapsed = round(self._timer.elapsed_seconds(), 2)
+    def report_ingestion_stage_start(self, stage: str) -> None:
+        if self._timer:
+            elapsed = self._timer.elapsed_seconds(digits=2)
             logger.info(
                 f"Time spent in stage <{self.ingestion_stage}>: {elapsed} seconds",
                 stacklevel=2,
             )
             if self.ingestion_stage:
                 self.ingestion_stage_durations[self.ingestion_stage] = elapsed
-
-    def report_ingestion_stage_start(self, stage: str) -> None:
-        self._close_stage()
+        else:
+            self._timer = PerfTimer()
 
         self.ingestion_stage = f"{stage} at {datetime.now(timezone.utc)}"
         logger.info(f"Stage started: {self.ingestion_stage}")
         self._timer.start()
 
-    def close_stage(self) -> None:
-        # just close ongoing stage if any
-        self._close_stage()
+
+@dataclass
+class IngestionStageContextReport:
+    ingestion_stage_durations: TopKDict[str, float] = field(default_factory=TopKDict)
+
+    def new_stage(self, stage: str) -> "IngestionStageContext":
```

Comment:
I'm a bit confused why this can't just be a method on the existing `IngestionStageReport` type?

Like I was thinking
1. add `IngestionStageReport.new_stage` method
2. deprecate + migrate all existing usages of `report_ingestion_stage_start` and `close_stage` to use the new_stage context manager variant
3. remove the deprecated methods

You can use the `@contextmanager` decorator to avoid needing a whole new class

Reply:
There is nothing preventing to add the method on the existing class. I opted for separate class to highlight the new implementation using context managers, incorporating "context" in the class name for clarity.

Also, I typically like to keep the scope of a PR focused, so I separate feature and refactor changes to make reviews easier.

Anyway, as suggested I kept existing class name and did refactor of all other usages of ingestion stage in commit [e7825e4](https://github.com/datahub-project/datahub/pull/12234/commits/e7825e4e9f379169ae7614e25fa0e0185f5eb87f)